eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs

```diff
@@ -632,14 +633,14 @@ modifyImpInitProtVer ver =
 
 modifyImpInitExpectLedgerRuleConformance ::
   forall era.
-  ShelleyEraImp era =>
-  ( Either
+  ( forall t.
+    Either
       (NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
       (State (EraRule "LEDGER" era), [Event (EraRule "LEDGER" era)]) ->
     LedgerEnv era ->
     LedgerState era ->
     Tx era ->
-    Expectation
+    ImpM t ()
```

Comment:
The original idea behind switching to using plain old `Expectation` was the following...

> Moreover, I am leaning towards making it an Expectation. This would limit access of this hook to the ImpTest state and protect against mistakes that would affect the outcome of the imp test itself.

https://github.com/IntersectMBO/cardano-ledger/pull/4748#discussion_r1849411456

Reply:
Yes, this also accomplishes the same goal, because as soon as you try to access the state, the typechecker will complain that you're not providing a general enough function.

Reply:
But now the upside is that you can still use functions that output an `ImpM t a` that don't modify the state (e.g. `logString`, `shouldBe`)

Reply:
That's really cool. @Soupstraw :+1:

Reply:
Could you add a comment to that effect near the definition, so that in the future if someone tried to do that, they would know that the function signature is fully polymorphic there for this reason?